src/modules/emit_sound.rs

```diff
@@ -0,0 +1,219 @@
+//! `bxt_emit_sound`
+
+use std::ffi::CString;
+use std::num::ParseFloatError;
+use std::ptr::NonNull;
+use std::str::FromStr;
+
+use super::Module;
+use crate::handler;
+use crate::hooks::engine;
+use crate::modules::commands::Command;
+use crate::utils::*;
+
+pub struct EmitSound;
+impl Module for EmitSound {
+    fn name(&self) -> &'static str {
+        "bxt_emit_sound"
+    }
+
+    fn description(&self) -> &'static str {
+        "Exposing some sound functions."
+    }
+
+    fn commands(&self) -> &'static [&'static Command] {
+        static COMMANDS: &[&Command] = &[
+            &BXT_EMIT_SOUND_STOP_SOUND_EXCEPT,
+            &BXT_EMIT_SOUND_DYNAMIC,
+            &BXT_EMIT_SOUND,
+        ];
+        COMMANDS
+    }
+
+    fn is_enabled(&self, marker: MainThreadMarker) -> bool {
+        engine::SV_StartSound.is_set(marker)
+            && engine::S_StartDynamicSound.is_set(marker)
+            && engine::S_StopSound.is_set(marker)
+            && engine::S_PrecacheSound.is_set(marker)
+    }
+}
+
+#[derive(Default)]
+struct SoundInfo {
+    sound: String,
+    channel: i32,
+    /// Volume is [0..1]
+    volume: f32,
+    from: i32,
+    to: i32,
+    /// This is the "radius" of the sound. Close to 1, you hear sound quieter when moving farther
+    /// from entity. "NORM" for sound-emitting entity is 0.8.
+    /// 0 is not quite to make it "global". Negative value will do.
+    attenuation: f32,
+    flag: i32,
+    /// 100 is no shift.
+    pitch: i32,
+}
+
+impl SoundInfo {
+    fn new(sound: String, channel: i32) -> Self {
+        SoundInfo {
+            sound,
+            channel,
+            volume: 1.,
+            from: 0,
+            to: 0,
+            attenuation: 0.,
+            flag: 0,
+            pitch: 100,
+        }
+    }
+}
+
+// Eh.
+// Parse float then round then convert to integer to avoid a more convoluted solution.
+impl FromStr for SoundInfo {
+    type Err = ParseFloatError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut rv = SoundInfo::default();
+
+        let mut iter = s.split_ascii_whitespace();
+        rv.sound = iter.next().unwrap_or_default().to_string();
+        rv.channel = (f32::from_str(iter.next().unwrap_or_default())?).round() as i32;
+        rv.volume = f32::from_str(iter.next().unwrap_or_default())?;
```

Comment:
Should this and perhaps other fields do clipping? Maybe at the point of invoking StartSound?

Reply:
I am not sure what this means.

Reply:
.clamp(0, 100) right here during parsing, or maybe at the point where SV_StartSound is called